appinventor/appengine/src/com/google/appinventor/client/TopToolbar.java

```diff
@@ -931,7 +940,239 @@ public void execute() {
+        } else if (json.type == 'logout') {
+          console.log("type is logout")
+          @com.google.appinventor.client.TopToolbar.InstantHelpAction::closeDialog(Lcom/google/gwt/user/client/ui/DialogBox;Ljava/lang/String;)(@com.google.appinventor.client.TopToolbar.InstantHelpAction::db, d);
+          @com.google.appinventor.client.TopToolbar.InstantHelpAction::logout()();
+        } else if (json.type == 'get_user_email') {
+          $doc.getElementById("AskHelpFrame").contentWindow.postMessage({
+            type: 'email',
+            username: userEmail
+          });
+        }
+      }
+    }-*/;
+
+    private static void submitPost(String username, String title, String description, int categoryId, boolean attachProject, String projectId) { // passing project id as string because long is not safe to access in JSNI
+      Ode.getInstance().getSubmitPostService().submitPost(Ode.getInstance().getUser().getUserId()
+      , username, title, description, categoryId
+      , attachProject, projectId, new AsyncCallback<String> (){ 
+          @Override
+          public void onSuccess(String response) {
+            Ode.CLog(response);
+            JSONObject jsonObject = JSONParser.parseStrict(response).isObject();
+            if (jsonObject != null) {
+              if (jsonObject.get("errors") == null || jsonObject.get("errors").isArray() == null) {
+                submitPostResult(false, null);
+              } else {
+                JSONArray errors = jsonObject.get("errors").isArray();
+                if (errors.get(0) != null)
+                  submitPostResult(true, errors.get(0).isString().stringValue());
+                else 
+                  submitPostResult(true, "Something went wrong.");
+              }
+            } else {
+              submitPostResult(true, "Something went wrong.");
+            }
+          }
+  
+          @Override
+          public void onFailure(Throwable reason) {
+            Ode.CLog("ERROR: " + reason.getMessage() + reason.getStackTrace()[0].getLineNumber());
+            submitPostResult(true, "Something went wrong.");
+          }
+        });
+    }
+
+    private static native void submitPostResult(boolean error, String errorMessage) /*-{
+      if (error) {
+        $doc.getElementById("AskHelpFrame").contentWindow.postMessage({
+          type: 'error',
+          message: errorMessage
+        });
+      } else {
+        $doc.getElementById("AskHelpFrame").contentWindow.postMessage({
+          type: 'upload_success',
+        });
+      }
+      
+    }-*/;
+
+    private static native void updateEmail(String username) /*-{
+      $doc.getElementById("AskHelpFrame").contentWindow.postMessage({
+        type: 'email',
+        username: username
+      });
+    }-*/;
+
+    private static void logout() {
```

Comment:
Might want to name this logoutFromCommunity or something similar so readers do not get confused and believe it logs you out from App Inventor itself.